File: JavaScript/WebsiteCode/LeechCode/Easy/DefuseTheBomb.js

// 0ms 50.39MB
var decrypt = function(code, k) {
    const n = code.length;
   const result = new Array(n).fill(0);

   if (k === 0) return result;

   let start, end, sum = 0;

   // Define start and end based on k
   if (k > 0) {
       start = 1;
       end = k;
   } else {
       start = n + k;
       end = n - 1;
   }

   // Initialize the sum for the first element
   for (let i = start; i <= end; i++) {
       sum += code[i % n];
   }

   // Apply sliding window to compute the sum for each position
   for (let i = 0; i < n; i++) {
       result[i] = sum;
       // Slide the window
       sum -= code[start % n];
       start++;
       end++;
       sum += code[end % n];
   }

   return result;
};

// 3ms 50.12MB
var decrypt = function(code, k) {
    const n = code.length;
    if (k === 0) return Array(n).fill(0);

    let isReversed = false;

    if (k < 0) {
        code.reverse();
        k = -k;
        isReversed = true;
    }

    const pre = new Array(n).fill(0);
    pre[0] = code[0];
    for (let i = 1; i < n; i++) {
        pre[i] = pre[i - 1] + code[i];
    }

    const res = new Array(n).fill(0);
    for (let i = 0; i < n; i++) {
        if (i + k < n) {
            res[i] = pre[i + k] - pre[i];
        } else {
            res[i] = (pre[n - 1] - pre[i]) + pre[(i + k) % n];
        }
    }

    if (isReversed) {
        res.reverse();
    }

    return res;
};

File: Python/WebsiteCode/LeetCode/Easy/Array/DefuseTheBomb.py
from ast import List

# 14ms 16.77MB
class Solution:
    def decrypt(self, a: List[int], k: int) -> List[int]:
        return [sum(a[(i+j)%len(a)] for j in range(*((k,0),(1,k+1))[k>0])) for i in range(len(a))]

# 0ms 16.71MB
class Solution:
    def decrypt(self, code: List[int], k: int) -> List[int]:
        n = len(code)  # O(1): Length of the list.
        
        if k == 0:
            return [0] * n  # O(n): Return a list of zeros if k is 0.

        result = [0] * n  # O(n): Initialize the result list.
        
        # Determine the direction of the sliding window.
        start, end, step = (1, k + 1, 1) if k > 0 else (k, 0, 1)
        window_sum = sum(code[i % n] for i in range(start, end))  # O(k): Initialize the first window sum.

        for i in range(n):  # O(n): Slide the window across all positions.
            result[i] = window_sum  # Store the current window sum in the result.
            # Slide the window: Remove the outgoing element and add the incoming one.
            window_sum -= code[(i + start) % n]  # O(1): Remove outgoing element.
            window_sum += code[(i + end) % n]    # O(1): Add incoming element.
        
        return result  # O(1): Return the result list.